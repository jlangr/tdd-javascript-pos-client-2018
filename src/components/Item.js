import React, { Component } from 'react';
import { Form, FormGroup, FormControl, ControlLabel, Button } from 'react-bootstrap';

class Item extends Component {
  constructor() {
    super();
    this.state = { id: 0, description: '', price: 0 };
  }
  render() { 
    return (
      <div className='item'>
        <Form>
          <FormGroup>
            <ControlLabel>Item</ControlLabel>
            <FormControl 
              className='input-description'
              onChange={event => this.setState({ description: event.target.value })} />
            <ControlLabel>Price</ControlLabel>
            <FormControl 
              className='input-price'
              onChange={event => this.setState({ price: event.target.value })} />
          </FormGroup>
        </Form>
        <Button 
          className='btn-void'
          onClick={() => this.props.voidItem(this.props.item.id)} >
          Void
         </Button>
      </div>
    );
  }
}

export default Item;