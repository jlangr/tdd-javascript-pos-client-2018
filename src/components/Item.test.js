import React from 'react';
import { shallow } from 'enzyme';
import Item from './Item';

describe('Item', () => {
  const mockVoid = jest.fn();
  const item = { id: 1, description: 'milk', price: 0 };
  const props = { item, voidItem: mockVoid };
  const itemComponent = shallow(<Item {...props}></Item>);

  it('renders properly', () => {
    expect(itemComponent).toMatchSnapshot();
  });

  it('initializes an item in `state`', () => {
    expect(itemComponent.state()).toEqual({ id: 0, description: '', price: 0 });
  });


  describe('when typing into the price input', () => {
    const price = 4.00;

    beforeEach(() => {
      itemComponent.find('.input-price').simulate('change', { target: { value: price }});
    });

    it('updates the price in `state`', () => {
      expect(itemComponent.state().price).toEqual(price);
    });
  });

  describe('when clicking `void item` button', () => {
    beforeEach(() => {
      itemComponent.find('.btn-void').simulate('click');
    });

    it('calls the void item callback', () => {
      expect(mockVoid).toHaveBeenCalledWith(item.id);
    });
  });
});