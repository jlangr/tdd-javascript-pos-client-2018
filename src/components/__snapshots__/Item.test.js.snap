// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Item
    item={
        Object {
            "description": "milk",
            "id": 1,
            "price": 0,
          }
    }
    voidItem={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Form
          bsClass="form"
          componentClass="form"
          horizontal={false}
          inline={false}
>
          <FormGroup
                    bsClass="form-group"
          >
                    <ControlLabel
                              bsClass="control-label"
                              srOnly={false}
                    >
                              Item
                    </ControlLabel>
                    <FormControl
                              bsClass="form-control"
                              className="input-description"
                              componentClass="input"
                              onChange={[Function]}
                    />
                    <ControlLabel
                              bsClass="control-label"
                              srOnly={false}
                    >
                              Price
                    </ControlLabel>
                    <FormControl
                              bsClass="form-control"
                              className="input-price"
                              componentClass="input"
                              onChange={[Function]}
                    />
          </FormGroup>
</Form>,
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          className="btn-void"
          disabled={false}
          onClick={[Function]}
>
          Void
</Button>,
      ],
      "className": "item",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bsClass": "form",
          "children": <FormGroup
            bsClass="form-group"
>
            <ControlLabel
                        bsClass="control-label"
                        srOnly={false}
            >
                        Item
            </ControlLabel>
            <FormControl
                        bsClass="form-control"
                        className="input-description"
                        componentClass="input"
                        onChange={[Function]}
            />
            <ControlLabel
                        bsClass="control-label"
                        srOnly={false}
            >
                        Price
            </ControlLabel>
            <FormControl
                        bsClass="form-control"
                        className="input-price"
                        componentClass="input"
                        onChange={[Function]}
            />
</FormGroup>,
          "componentClass": "form",
          "horizontal": false,
          "inline": false,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-group",
            "children": Array [
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
>
                Item
</ControlLabel>,
              <FormControl
                bsClass="form-control"
                className="input-description"
                componentClass="input"
                onChange={[Function]}
/>,
              <ControlLabel
                bsClass="control-label"
                srOnly={false}
>
                Price
</ControlLabel>,
              <FormControl
                bsClass="form-control"
                className="input-price"
                componentClass="input"
                onChange={[Function]}
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "control-label",
                "children": "Item",
                "srOnly": false,
              },
              "ref": null,
              "rendered": "Item",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "form-control",
                "className": "input-description",
                "componentClass": "input",
                "onChange": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "control-label",
                "children": "Price",
                "srOnly": false,
              },
              "ref": null,
              "rendered": "Price",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "form-control",
                "className": "input-price",
                "componentClass": "input",
                "onChange": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "active": false,
          "block": false,
          "bsClass": "btn",
          "bsStyle": "default",
          "children": "Void",
          "className": "btn-void",
          "disabled": false,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Void",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Form
            bsClass="form"
            componentClass="form"
            horizontal={false}
            inline={false}
>
            <FormGroup
                        bsClass="form-group"
            >
                        <ControlLabel
                                    bsClass="control-label"
                                    srOnly={false}
                        >
                                    Item
                        </ControlLabel>
                        <FormControl
                                    bsClass="form-control"
                                    className="input-description"
                                    componentClass="input"
                                    onChange={[Function]}
                        />
                        <ControlLabel
                                    bsClass="control-label"
                                    srOnly={false}
                        >
                                    Price
                        </ControlLabel>
                        <FormControl
                                    bsClass="form-control"
                                    className="input-price"
                                    componentClass="input"
                                    onChange={[Function]}
                        />
            </FormGroup>
</Form>,
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="btn-void"
            disabled={false}
            onClick={[Function]}
>
            Void
</Button>,
        ],
        "className": "item",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form",
            "children": <FormGroup
              bsClass="form-group"
>
              <ControlLabel
                            bsClass="control-label"
                            srOnly={false}
              >
                            Item
              </ControlLabel>
              <FormControl
                            bsClass="form-control"
                            className="input-description"
                            componentClass="input"
                            onChange={[Function]}
              />
              <ControlLabel
                            bsClass="control-label"
                            srOnly={false}
              >
                            Price
              </ControlLabel>
              <FormControl
                            bsClass="form-control"
                            className="input-price"
                            componentClass="input"
                            onChange={[Function]}
              />
</FormGroup>,
            "componentClass": "form",
            "horizontal": false,
            "inline": false,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-group",
              "children": Array [
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}
>
                  Item
</ControlLabel>,
                <FormControl
                  bsClass="form-control"
                  className="input-description"
                  componentClass="input"
                  onChange={[Function]}
/>,
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}
>
                  Price
</ControlLabel>,
                <FormControl
                  bsClass="form-control"
                  className="input-price"
                  componentClass="input"
                  onChange={[Function]}
/>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "control-label",
                  "children": "Item",
                  "srOnly": false,
                },
                "ref": null,
                "rendered": "Item",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "form-control",
                  "className": "input-description",
                  "componentClass": "input",
                  "onChange": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "control-label",
                  "children": "Price",
                  "srOnly": false,
                },
                "ref": null,
                "rendered": "Price",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "form-control",
                  "className": "input-price",
                  "componentClass": "input",
                  "onChange": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": false,
            "block": false,
            "bsClass": "btn",
            "bsStyle": "default",
            "children": "Void",
            "className": "btn-void",
            "disabled": false,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Void",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "disableLifecycleMethods": true,
    "hydrateIn": undefined,
  },
}
`;
